### ISSUE TITLE
Multipass Launch timed out waiting for response on Windows 10

### ISSUE THREAD

**Issue by robogeek** [canonical/multipass#1512]
**Describe the bug**

I'm trying to replicate on Windows what I do on macOS X.  I have a fresh install of Multipass 1.2.1, and try to run "multipass launch", and I get this:

```
PS C:\Users\heavy> multipass launch
launch failed: The following errors occurred:
unassuming-tailorbird: timed out waiting for response
```

**To Reproduce**

1. `multipass launch`
1. Long pause during which time various messages about _Creating_ and _Verifying_ are printed
1. Eventually the above message is printed

**Expected behavior**

I expect a Multipass instance to launch

**Logs**

I tried running `multipass launch -v` but no extra anything was printed.

In the Event Viewer (as suggested in the docs) I don't see anything

```
[strapping-diver] cmdlet: 'Get-VM -Name strapping-diver \| Select-Object  -ExpandProperty State'
--
```

**Additional info**
 - OS: Windows 10 Pro Insider Preview - Version 2004 - Installed on 4/27/2020 - OS Build 19612.1005 - Windows Feature Experience Pack 120.8400.0.0
- Multipass version 1.2.1

```
PS C:\Users\heavy> multipass info --all
info failed: ssh connection failed: 'Failed to resolve hostname dazzled-bittern.mshome.net (No such host is known. )'
```

**Additional context**

I have WSL2 installed - but I have disabled it by going into _Windows Features_ and unchecking the WSL2 checkbox.   I did that because of comments about WSL2 being incompatible with Multipass because of how WSL2 takes over Hyper-V.

For every attempted `multipass launch` there is a machine listed in Hyper-V Manager.   If I attempt to Connect with the machine using the Manager app, it launches a window that ends up containing this message:

```
Error: invalid arch-independend ELF magic.
Entering rescue mode...
grub rescue>
```

I don't know what to do at that point, and _help_ doesn't give anything other than an error.

The _multipass list_ command shows all the attempts listed in the Running state, and the Hyper-V manager also shows them as Running.

I suppose another bit of context is that I despise Windows, and this experience is only validating my despite for Windows.  But the more practical thing is that I barely understand what to do.

The last bit is that I am working on the 5th Edition of my book, _Node.js Web Development_, and am hoping to showcase Multipass as a good solution for testing deployment to a virtual server on ones laptop.  It works fine on MacOSX but I'm of course having to verify that everything works on Windows.  At this point I don't know what to do.  Maybe the things I want the reader to do will work on WSL2 instead.

**Comment by alanisaac** [comment]
I hit this exact same issue today.  Multipass was working fine for a few weeks on my Windows machine, but mysteriously stopped launching VMs with the exact same errors above.  I tried all the steps above and saw the same results (except WSL2, which I don't have enabled in the first place).

My system is:
```
multipass  1.1.0+win
Windows 10 Enterprise, Version 1903, OS build 18362.815
```

**Comment by jfmatth** [comment]
Does the vm you're building get an IP from the command:
```
multipass list
```

I recently had some issues after i upgraded from 1.1 -> 1.2 -> 1.21, nothing worked.  I uninstalled and re-installed.

**Comment by robogeek** [comment]
No, there is no IP address shown.

Since I had also done upgrades to get to 1.21, I tried the following:

* With WSL2 shut down (`wsl.exe --shutdown`)
    * Multipass still misbehaves w/ the timeouts shown above
* Uninstall Multipass, telling it to not delete the existing instances
* In Hyper-V, stop each VM then delete each VM - Hyper-V shows no VM's existing
* Install Multipass
* On running `multipass list` the VM's become listed again in Hyper-V, and the `multipass list` output shows them in Stopped state
* Was able to readily `delete` then `purge` the old VM's, and `list` showed them as missing

I tried:  `multipass launch` again, and got the same timeout behavior.  At that point most Multipass commands would take a long time leading to the timeout.

I rebooted the system.

The VM I attempted to launch was in Stopped state, and `delete` and `purge` readily got rid of it.

Attempting to run `multipass launch` again resulted in the same timeout behavior.

**Comment by jfmatth** [comment]
i suspect it's an IP assignment issue, i'm not sure 

I had nothing but issues with all non-WSL2 products in the preview version of Windows, ended up uninstalling and reinstalling Windows.  I don't think WSL2 knows enough about networks it doesn't manage IMHO

is there a way you can remove WSL2, make sure it's networks disappear and try again?

There are some troubleshooting articles around HyperV and Multipass you might have a look at

**Comment by robogeek** [comment]
I have disabled WSL, which caused a reboot.  It doesn't show up as a regular application that can be uninstalled.  So far no luck however.

**Comment by jfmatth** [comment]
WSL is a Windows Feature you have to remove from there :(

I'm sorry you're having such issues, i love my setup so much i've written a repo around it and some scripts to make my life easier.  Check my Github repos.

- Check that the Multipass VM is connected to the default switch
- Do you have other Hyper-V virtual switches?
- Can you make regular VM's on Hyper-V?

Maybe w/o WSL installed, remove MP one more time and then reinstall?

**Comment by robogeek** [comment]
Until last night I was not aware that one could install a normal virtual machine using Hyper-V ... but then I saw this thing about Quick Create and voila I could launch a normal Ubuntu instance that way.

So yes I can create a regular VM on Hyper-V.  

As for virtual switches etc, I have no idea what you're talking about.  Presumably that's some kind of software emulated network switch.  Presumably that means an emulated ethernet hub.  But what I don't know is anything about how that's done in Hyper-V.

**Comment by robogeek** [comment]
I have a bit more of a clue as to what the virtual switch thing is, after watching a couple YouTube videos.

There's only one switch - the default switch - and it is configured as Internal.  I wasn't able to figure out how to change any of its settings, because the buttons are all grey'd out.

I have disabled the Windows Subsystem for Linux feature, the Hyper-V features are enabled.  Multipass is uninstalled, and I rebooted after doing that.  Hyper-V shows two VM's, DockerDesktopVM and "primary".  Both are Off, and if I recall right Docker is disabled currently.

**Comment by robogeek** [comment]
I have now installed Multipass and was unable to find any changes in any settings in the Hyper-V Manager application.

But..  I got this:

```
PS C:\Users\heavy> multipass launch --name focal
launch failed: The following errors occurred:
focal: timed out waiting for response
```

The VM named `focal` is shown in a Running state with 1gb of memory.  The network adapter is Default Switch - obviously since there's only that one switch.  "Enable virtual LAN identification" is not selected.  Connecting to it results in the same "grub rescue" screen discussed above.

In other words, the exact same behavior reported at the top.

**Comment by robogeek** [comment]
@jfmatth  you said there are troubleshooting articles.  Where?

**Comment by robogeek** [comment]
After searching (DuckDuckGo) for "troubleshoot hyper-v multipass" I found this:  https://multipass.run/docs/troubleshooting-networking-on-windows

It suggested 

```
Get-HNSNetwork | ? Name -Like "Default Switch" | Remove-HNSNetwork
```

So I did that, and rebooted.  After reinstalling Multipass again, I got this:

```
PS C:\Users\heavy> multipass launch --name focal
launch failed: The following errors occurred:
focal: timed out waiting for response
```

Again.  So frustrating since the behavior looked so good - it was readily downloading the image and configuration - only to result in this.  I tried again using a different machine name, in case there was something about reusing the same machine name again and again.  

FWIW - the tracing output got to "Starting" before the timeout message.

And... launching with "-v" (even up to "-v -v -v") did not give any more information.

**Comment by Saviq** [comment]
Hi @robogeek, try using a different name - or skip `--name` altogether. Window's ICS is known for getting confused with multiple (different) hosts using the same name too often.

Have a look at `C:\WINDOWS\System32\drivers\etc\hosts.ics` - a common case is that there will be multiple entries for the same hostname. You can either edit that file or remove it altogether.

**Comment by Saviq** [comment]
And thanks @jfmatth for trying to help out!

**Comment by jfmatth** [comment]
> And thanks @jfmatth for trying to help out!

Anytime, we've all been there.

MP is one of my favorite tools these days, and Hyper-V can be someone fussy at times.

**Comment by jfmatth** [comment]
@robogeek Thanks for hanging in there.

It seems like Multipass is working OK.  The behavior you're describing sounds like the MP VM isn't getting an IP in a timely manner, and I"m not sure what causes that, but lets try something else.

Back to making a VM with just Hyper-V, when you did that, did that VM get an IP? (```ip a``` )from within the Linux VM).  Make sure that VM is attached to the default switch.

how many network cards do you have on your system?  From your windows command prompt, do 
```
ipconfig /all
```

**Don't** paste the output there, but how many Ethernet adapters do you have?  What I found with the Preview version, it left ghost adapters behind, and that caused all sorts of issues.

One last thing to check, the Event Viewer might show some warnings or errors for MP, check that.

As I mentioned to Saviq above, Multipass is such an amazing tool, and Hyper-V is one of the best hypervisors, but it can be fussy.

Good luck, and we'll try to get you going

**Comment by jfmatth** [comment]
> Have a look at `C:\WINDOWS\System32\drivers\etc\hosts.ics` - a common case is that there will be multiple entries for the same hostname. You can either edit that file or remove it altogether.
> 

I just had issues accessing my primary instance, and this was the issue.  I deleted my host.ics file and rebooted, Hyper-V cleaned up the file and my primary was accessible again.

this might help @robogeek

**Comment by robogeek** [comment]
When I looked at `host.ics` yesterday there were two entries.  One looked to be well formed, but the other was just a pair of numbers on a line.  Immediately following that was a commented out entry that looked well formed.  I didn't change anything because of all the warnings in the file to not touch anything.

* With Multipass uninstalled
* Deleted `hosts.ics` and rebooted
* The `hosts.ics` was not recreated until I launched the normal Ubuntu VM that I discuss farther down.
* The resulting `hosts.ics` has two entries -- one for the laptop - the other for the normal Ubuntu VM instance - both are well formed - no spurious extra text
* Prior to this reboot, Multipass was installed, and `hosts.ics` had some spurious extra text

My system is a laptop - hence there is two interfaces - a wired ethernet and an M.2 WiFi card.  I'm not using the WiFi and from `ipconfig /all` it seems it isn't connected to WiFi at all.

As for installing a normal VM... I started to install a normal Ubuntu VM last night.  That finished overnight, and I just checked it out.

* The Hyper-V console does not show an IP address assigned.
* However, inside the VM it has an IP address assignment - 172.18.2.107
* And in the Windows host I can use `<host-name>.mshome.net` to not only ping the instance, but to use SSH to log in to the instance.
* And inside the instance it is readily able to access the Internet, such as to download packages to install and to update the system

Bottom line ... I now have a working system, thanks guys.

As noted in the section for `hosts.ics`, I had uninstalled Multipass, deleted `hosts.ics` and rebooted.  Upon reboot `hosts.ics` was not there.  Once I booted the regular Ubuntu VM, `hosts.ics` was created and it only had well formed entries.

At that point I thought to install Multipass again.  I've done that, and then launched an instance using no options - just the default name.  The instance came right up, and was assigned an IP address.  Okay, cool.  I launched another instance, this time specifying a name, and it too came right up and was assigned an IP address.

The `hosts.ics` file does not have any spurious text, instead it only has well formed entries.

That may be the issue - that my `hosts.ics` had spurious extra text.

**Comment by Saviq** [comment]
Perfect @robogeek! Thanks for the detailed report, too :)

**Comment by xuhu55** [comment]
Deleting hosts.ics does fix the issue temporarily for me but the issue keeps re-occuring every other time I try to make a new server. Is there a way to prevent hosts.ics from being corrupted?

**Comment by Saviq** [comment]
@xuhu55 does it get _corrupted_, or does it include double entries for the same name? If it's the latter, using unique instance names should be enough to avoid the problems.

**Comment by xuhu55** [comment]
> @xuhu55 does it get _corrupted_, or does it include double entries for the same name? If it's the latter, using unique instance names should be enough to avoid the problems.

It includes a line with just a number so it's corrupted and not a problem with uniqueness.

**Comment by jbergens** [comment]
I just got this error on my Windows 10 laptop. I can't see any problem in my hosts.ics, it only has one row and that looks correct. I have WSL2 and Docker activated but thought that should be supported.

**Comment by obok** [comment]
Mine `hosts.ics` was _corrupted_ as in @robogeek and @xuhu55 cases, i.e. it had lines with one or few numbers (space delimited). After manually deleting corrupted lines from `hosts.ics` (no restart) the `multipass` works properly!

Thanks @Saviq for the hint!

An open question is: what corrupts `hosts.ics`?

**Comment by Saviq** [comment]
@obok the `ics` part stands for Internet Connection Sharing. The built-in Windows feature, so it's a bug in the DHCP server in Windows as far we can tell.

**Comment by Saviq** [comment]
@killerdroid99 please open a new issue with the details requested in the issue template:

https://github.com/canonical/multipass/issues/new/choose
